driver-core/src/main/java/com/datastax/driver/core/AbstractAddressableByIndexData.java

```diff
@@ -119,117 +160,95 @@ public T setBytesUnsafe(int i, ByteBuffer v) {
     }
 
     public T setVarint(int i, BigInteger v) {
-        checkType(i, DataType.Name.VARINT);
-        return setValue(i, v == null ? null : TypeCodec.BigIntegerCodec.instance.serialize(v));
+        checkType(i, VARINT);
+        return setValue(i, v == null ? null : codecFor(i, BigInteger.class).serialize(v, protocolVersion));
     }
 
     public T setDecimal(int i, BigDecimal v) {
-        checkType(i, DataType.Name.DECIMAL);
-        return setValue(i, v == null ? null : TypeCodec.DecimalCodec.instance.serialize(v));
+        checkType(i, DECIMAL);
+        return setValue(i, v == null ? null : codecFor(i, BigDecimal.class).serialize(v, protocolVersion));
     }
 
     public T setUUID(int i, UUID v) {
-        DataType.Name type = checkType(i, DataType.Name.UUID, DataType.Name.TIMEUUID);
-
-        if (v == null)
-            return setValue(i, null);
-
-        if (type == DataType.Name.TIMEUUID && v.version() != 1)
-            throw new InvalidTypeException(String.format("%s is not a Type 1 (time-based) UUID", v));
-
-        return type == DataType.Name.UUID
-             ? setValue(i, TypeCodec.UUIDCodec.instance.serialize(v))
-             : setValue(i, TypeCodec.TimeUUIDCodec.instance.serialize(v));
+        checkType(i, UUID, TIMEUUID);
+        return setValue(i, v == null ? null : codecFor(i, UUID.class).serialize(v, protocolVersion));
     }
 
     public T setInet(int i, InetAddress v) {
-        checkType(i, DataType.Name.INET);
-        return setValue(i, v == null ? null : TypeCodec.InetCodec.instance.serialize(v));
+        checkType(i, INET);
+        return setValue(i, v == null ? null : codecFor(i, InetAddress.class).serialize(v, protocolVersion));
     }
 
+    @SuppressWarnings("unchecked")
     public <E> T setList(int i, List<E> v) {
-        DataType type = getType(i);
-        if (type.getName() != DataType.Name.LIST)
-            throw new InvalidTypeException(String.format("Column %s is of type %s, cannot set to a list", getName(i), type));
-
+        checkType(i, LIST);
         if (v == null)
             return setValue(i, null);
-
-        // If the list is empty, it will never fail validation, but otherwise we should check the list given if of the right type
-        if (!v.isEmpty()) {
-            // Ugly? Yes
-            Class<?> providedClass = v.get(0).getClass();
```

Comment:
Yes, absolutely. If a suitable codec is not found, the `CodecNotFoundException` would be thrown. Otherwise. if the returned codec encounters a problem, a `RuntimeException` would be thrown (most likely, `ClassCastException` or `InvalidTypeException`). Note that this could be a problem for clients that already catch specific unchecked exceptions such as `InvalidTypeException`, as the new implementation does not guarantee that the very same kind of exception will be raised. This should maybe be documented.